single_kernel_mongo/core/operator.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Abstract Operator for Mongo Related Charms."""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, ClassVar
+
+from ops.charm import RelationDepartedEvent
+from ops.framework import Object
+from ops.model import Unit
+
+from single_kernel_mongo.config.literals import RoleEnum, Substrates
+from single_kernel_mongo.managers.config import CommonConfigManager
+from single_kernel_mongo.managers.mongo import MongoManager
+from single_kernel_mongo.state.charm_state import CharmState
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.abstract_charm import AbstractMongoCharm
+    from single_kernel_mongo.core.workload import WorkloadBase
+    from single_kernel_mongo.events.database import DatabaseEventsHandler
+    from single_kernel_mongo.events.tls import TLSEventsHandler
+    from single_kernel_mongo.managers.tls import TLSManager
+
+
+class OperatorProtocol(ABC, Object):
```

Comment:
It will be easier to see when I have developed the second Operator` I think (which lives in another branch currently as I'm working on it. Let's talk about it this afternoon until I have some time to draw the charts and write documentation.